feat(ui): Position UI windows by anchor and offsets

al-ce opened the pull request:
Adds four new options to customize the position of the UI popup window(s):

- `vertical_anchor`
  - Options: `"top"`, `"center"`, `"bottom"`
  - Default: `"center"`
- `horizontal_anchor`
  - Options: `"left"`, `"center"`, `"right"`
  - Default: `"center"`
- `vertical_offset`
  - Default: 0
- `horizontal_offset`
  - Default: 0

The anchor and the offset values adjust the row and column position for each window (preview, patterns, replacements, options).

If the window would be out of bounds due to a combination of the anchor or offset opts and the overall size of the windows, the adjusted row/col positions of the UI are set to a max limit, placing them at their opposite anchor point.

al-ce committed:
feat(ui): Position UI windows by anchor and offsets

Adds four new options to customize the position of the UI popup window(s):

- `vertical_anchor`
  - Options: `"top"`, `"center"`, `"bottom"`
  - Default: `"center"`
- `horizontal_anchor`
  - Options: `"left"`, `"center"`, `"right"`
  - Default: `"center"`
- `vertical_offset`
  - Default: 0
- `horizontal_offset`
  - Default: 0

The anchor and the offset values adjust the row and column position for each window (preview, patterns, replacements, options).

If the window would be out of bounds due to a combination of the anchor or offset opts and the overall size of the windows, the adjusted row/col positions of the UI are set to a max limit, placing them at their opposite anchor point.

al-ce commented:
Thanks for the opportunity to collaborate! I leaned towards making the options and the function a bit blunt with their naming and structure but I'm happy to do any refactors you'd like to see. Please push any changes you see fit to make.

al-ce commented on lua/muren/ui.lua:
```diff
@@ -412,14 +460,13 @@ M.open = function(opts)
 
-  local gheight, gwidth = get_nvim_ui_size()
-
+  local positions = make_ui_positions()
```
(forgot to change this to match the other local var)

I think we could only call this once and pass it around as an argument to the function calls that need it, but I thought there might be some advantage to calling it multiple times that's not obvious, maybe if the values need to be updated when the ui window gets moved during resizing, etc.

AckslD replied:
Yeah I think indeed it's good to compute every time exactly since the window size may change in between :+1:

AckslD commented:
@al-ce seems really cool! I'm a bit swamped with things atm so sorry for not getting to this yet but just wanted to let you know it's still on my todo :)

al-ce commented:
> @al-ce seems really cool! I'm a bit swamped with things atm so sorry for not getting to this yet but just wanted to let you know it's still on my todo :)

No worries, take your time! Hope all is well

AckslD commented on lua/muren/ui.lua:
```diff
@@ -249,17 +249,65 @@ local get_nvim_ui_size = function()
+    },
+  }
+  local adjusted_row = positions.row[vertical_anchor] + vertical_offset
+  local adjusted_col = positions.col[horizontal_anchor] + horizontal_offset
```
do you think it would make sense if eg `horizontal_anchor = 'right'` then the offset is negative instead such that the offset is from the anchor?

al-ce replied:
(writing this with an eye to the nui discussion in #24 ) do you mean something like, change the calculation so the user setting for an offset is always a positive value? e.g. if h_anchor is 'right', then subtract the (positive) value of the h_offset, else if h_anchor is 'left', add it? (and similarly for v_anchor). This way the offset is always referencing how far off it is relative the anchor instead of being a shift relative to the screen as a whole.

Or did you mean something else?

AckslD replied:
Yes exactly :)

AckslD commented:
Could also be done in a follow up PR but do you think it would be useful to pass eg the anchor specification as an argument to the command? Such that you can open it where you want without changing the config?

al-ce committed:
refactor(ui): Calculate offsets relative to anchor, not screen

al-ce committed:
refactor(ui): Improve readability of make_ui_positions()

- Set ui position defaults in make_ui_positions(), not in setup
- Pass populated opts table as arg to make_ui_positions (rather than assuming opts table is in scope)
- Set default ui positioning to center if opts.vertical_anchor or opts.horizontal_anchor is not "top/bottom/left/right", rather than explicitly setting "center". This also prevents raising an error if an invalid value is set (e.g. "centered" etc.)
- Refactor 'out of bounds' calculations using more concise logic with math.max/math.min
- Rename ambiguous 'positions' table to 'anchors' and remove unnecessary nested tables
- Restructure returned table, grouping row/col for each sub-window
- Update README accordingly

al-ce committed:
feat(commands): Handle arguments for ui positioning

- Allow user to pass anchor opts through lua api calls. e.g. `api.open_ui({anchor = "top_left", vertical_offset = "5", horizontal_offset = "5", range = 2})` or commands (:MurenOpen top_left 5 2). This lets the user position the window via the args without having to change the opts.
- Condense opts.vertical_anchor and opts.horizontal_anchor into opts.anchor and use string.match to set the opt accordingly. So long as anchor contains "top" or "bottom" ('top' is matched first) or "left" or "right" ('left' is matched first), the corresponding anchor will be set. README is updated to recommend, e.g., "top", "top-left", "bottom-right", etc.
- Move get_range function from api.lua to options.lua
- Set default anchor and offset opts in options.lua
- Notify user if anchor value is invalid and if so set to 'center', open Muren in default position
- Create default muren commands with anchor-subcommand completion (e.g. `MurenOpen top`, `MurenToggle bottom_left` etc.) for first arg only. Second and third args (vertical and horizontal offset) do not show a completion popup, since they are intended to be integers.
- Update README accordingly

al-ce commented:
> Could also be done in a follow up PR but do you think it would be useful to pass eg the anchor specification as an argument to the command? Such that you can open it where you want without changing the config?

I tackled this in a different branch. https://github.com/al-ce/muren.nvim/commit/8e5cc6dbe250eec159168d5cbc447d08e133c76b 

https://github.com/al-ce/muren.nvim/assets/23170004/66d40c36-3265-44f0-88e4-4712fc00e957

EDIT: The video above doesn't show it but the subcommands respect the preset offset options (which were set to 0 for this demo). Could increase `nargs` to accept offsets here too (I think?)

https://github.com/al-ce/muren.nvim/assets/23170004/b5ea8dec-1ea0-4eca-863f-03ac0bc49966



It touches code in more modules than this PR currently does and makes some stronger design choices so I wanted to check with you before pushing it here. Moreover, if we decide to abandon this PR for a `nui.nvim` redesign (or even if we go ahead with this but eventually replace it), you probably don't want to deal with reworking those changes in that other branch. Just posting it here as a reference!

If you prefer the style of 4792b69 over the refactors in c669a6f, they can be reverted.

Thanks again for the opportunity, this is a fun weekend exercise! No worries if we don't merge this given the discussion around `nui`, which we can continue in the open issue. I just didn't want to leave the comments you left unaddressed. Hope all is well!

AckslD commented:
Hi @al-ce, I think both this PR and the changes for the commands look good. We can either merge this and then the updates to the commands or you first merge the updates to the commands here. Up to you what you find easier.

AckslD reviewed (approved):
Very cool :)

al-ce commented:
Thanks! We should probably go ahead and merge it here since there would be a breaking change from this PR's latest commit to the ones from the other branch. In both the Lua function args and the command args, the anchor is set with some combination of top/bottom + left/right. So I changed the user-end opts to reflect that, condensing vertical_anchor and horizontal_anchor, to `anchor`, which can be one of:

'top', 'bottom', 'left', 'right', 'top_left', 'top_right', 'bottom_left', 'bottom_right'

If the value isn't one of these, the anchor value is set to center, we notify the user without raising errors, and open muren as it would by default (rather than raising an error and preventing muren from functioning normally, since ui position is a secondary concern). The video below shows my setup.

I also added second and third position args for the :Muren commands. For example, `:MurenOpen`, `:MurenOpen top`, `:MurenOpen top 15`, `:MurenOpen top 15 5` are all valid. The last two are vertical offset and horizontal offset, which will override the default or the user's offset opts. Otherwise those opts are respected.

https://github.com/al-ce/muren.nvim/assets/23170004/029db738-1a49-4ba6-a683-efd3f9585f56

Thanks again for taking a look at this!

AckslD commented on lua/muren.lua:
```diff
@@ -1,17 +1,54 @@
+      vim.api.nvim_create_user_command(name, function(args)
+        local anchor, vertical_offset, horizontal_offset = unpack(args.fargs)
+        muren_api({
+          anchor = anchor,
```
```suggestion
          range = args.range,
          anchor = anchor,
```
such that substitution only in a visual selection works

al-ce replied:
Thanks, sorry I overlooked that! Too hasty on the refactor.

al-ce replied:
I made some fixes but I want to sleep on them and get a fresh look tomorrow! Main thing is that for the `get_range` function we also need line1 and line2 from the `args` here.
```suggestion
          range = args.range,
          line1 = args.line1,
          line2 = args.line2,
          anchor = anchor,
```

I'm also trying to think of some good-looking way to let a direct call to the api get the current visual selection, since e.g. `require("muren.api").open_ui({some_opts_here})` won't have the visual selection. My current hack and slash approach is to modify get_range like this:

```lua
local get_range = function(opts)
  local range
  if opts.range == 2 or vim.api.nvim_get_mode().mode == 'V' then
    local start_line, end_line = vim.fn.getpos("v")[2], vim.fn.getpos(".")[2]
    range = {start = opts.line1 or start_line, _end = opts.line2 or end_line}
  end
  return range
end
```

But I'll work on cleaning that up

al-ce committed:
fix(commands): Pass range opts to command for v-mode substitutions

al-ce committed:
refactor: Move get_range def back to api module

- Move create_muren_commands def out of setup func scope
- Remove 'range' bool from command creation, j
- Ensure MurenUnique gets a table value

al-ce commented:
Pushed one commit to address the visual selection issue and another to clean up some of the mess I made in prior commits 😅

AckslD commented:
Thanks for all the work @al-ce, high quality of code and great communication! :)

al-ce commented:
Thanks for the opportunity and the guidance! Enjoyed it
